idaes/power_generation/carbon_capture/mea_solvent_system/property_package/liquid_prop.py

```diff
@@ -0,0 +1,1282 @@
+                               domain=NonNegativeReals,
+                               initialize=10,
+                               doc="concentration of true species "
+                               "[mol/m3]")
+        def rule_logC(blk,i):
+          return log(blk.conc_comp_true[i])
+
+        self.logC = Expression(self._params.component_list_true,
+                               rule = rule_logC,
+                               doc='logarithm of true species concentration')
+
+        def rule_speciation_model(blk):
+          xa  = blk.conc_comp
+          xt  = blk.conc_comp_true
+          CT  = blk.logC
+          lnk = blk.lnk_eq_conc
+
+          #chemical equilibruim reaction 1
+          yield lnk['R1'] == \
+               CT['MEA+'] + CT['MEACOO-'] - 2*CT['MEA'] - CT['CO2']
+          #chemical equilibruim reaction 2
+          yield lnk['R2'] == \
+               CT['MEA+'] + CT['HCO3-'] - CT['MEA'] -  CT['CO2'] - CT['H2O']
+          #MEA balance
+          yield xa['MEA'] ==  xt['MEA'] + xt['MEA+'] +  xt['MEACOO-']
+          #CO2 balance
+          yield xa['CO2'] ==  xt['CO2'] + xt['HCO3-'] + xt['MEACOO-']
+          #H2O balance
+          yield xa['H2O'] ==  xt['H2O'] + xt['HCO3-']
+          #charge balance
+          yield xt['MEA+'] == xt['HCO3-'] + xt['MEACOO-']
+
+        try:
+            # Try to build speciation model constraints
+            self.speciation_model = ConstraintList(rule=rule_speciation_model)
+        except AttributeError:
+            # If constraint fails, clean up so that DAE can try again later
+            self.del_component(self.conc_comp_true)
+            self.del_component(self.logC)
+            self.del_component(self.speciation_model)
+            raise
+
+    def _make_state_vars(self):
+
+        self.flow_mol = Var(initialize=1.0,
+                            domain=NonNegativeReals,
+                            doc='Total molar flowrate [mol/s]')
+        self.mole_frac = Var(self._params.component_list,
+                             domain=NonNegativeReals,
+                             bounds=(0, 1),
+                      initialize=1 / len(self._params.component_list),
+                      doc='Component mole fractions [-]')
+        self.pressure = Var(initialize=101325,
+                            domain=NonNegativeReals,
+                            doc='Pressure [Pa]')
+        self.temperature = Var(initialize=298.15,
+                               domain=NonNegativeReals,
+                               doc='Temperature [K]')
+
+    def _flow_mol_comp(self):
+        self.flow_mol_comp = Var(self._params.component_list,
+                    initialize=1.0,
+                    domain=NonNegativeReals,
+                    doc='Component molar flowrate [mol/s]')
+
+        def rule_flow_mol_comp(b,j):
+            return  b.flow_mol_comp[j] == b.mole_frac[j]*b.flow_mol
+
+        self.flow_mol_comp_eq = Constraint(self._params.component_list,
```

Comment:
This is where you need the "clean-up for DAE" checks.

Due to the way the DAE transformation works, it is possible for this to be called before the state vars have been created (i.e. `flow_mol`). If this happens, an `AttributeError` is raised which causes the DAE transformation to push whatever called for this to the back of the queue and to try again later.

However, Pyomo does not clean up after itself when something fails to build - the `flow_mol_comp_eq` Constraint object now exists, but is empty (as it failed to build the necessary expression), as well as the `flow_mol_comp` Var. When the DAE transformation comes to try to build whatever triggered this call again, it sees the `flow_mol_comp` Var and thinks everything is good, even though the Constraint is actually empty.

Thus, we need to add the clean-up steps here so that if the constraint fails to build, we delete the empty constraint object AND any Vars we built so that when the DAE transformation come back it has a clean slate to work with.

Reply:
done.
```
    def rule_flow_mol_comp(b, j):
      return b.flow_mol_comp[j] == b.mole_frac[j] * b.flow_mol

    try:
        self.flow_mol_comp = Var(self._params.component_list,
                                 initialize=1.0,
                                 domain=NonNegativeReals,
                                 doc='Component molar flowrate [mol/s]')
        self.flow_mol_comp_eq = Constraint(self._params.component_list,
                                       rule=rule_flow_mol_comp,
                                       doc="Component molar flow in liquid phase"
                                       " [mol/s]")
    except AttributeError:
      # If constraint fails, clean up so that DAE can try again later
      self.del_component(self.flow_mol_comp)
      self.del_component(self.flow_mol_comp_eq)
      raise

```